src/main/java/io/pravega/connectors/flink/FlinkPravegaWriter.java

```diff
@@ -117,48 +164,411 @@ public void setPravegaWriterMode(PravegaWriterMode writerMode) {
+        private final long checkpointId;
+
+        TransactionAndCheckpoint(Transaction<T> transaction, long checkpointId) {
+            this.transaction = transaction;
+            this.checkpointId = checkpointId;
+        }
+
+        Transaction<T> transaction() {
+            return transaction;
+        }
+
+        long checkpointId() {
+            return checkpointId;
+        }
+
+        @Override
+        public String toString() {
+            return "(checkpoint: " + checkpointId + ", transaction: " + transaction.getTxnId() + ')';
+        }
+    }
+
+    /*
+     * An abstract internal writer.
+     */
+    private abstract class AbstractInternalWriter {
+
+        protected final EventStreamWriter<T> pravegaWriter;
+
+        AbstractInternalWriter(ClientFactory clientFactory) {
+            Serializer<T> eventSerializer = new FlinkSerializer<>(serializationSchema);
+            EventWriterConfig writerConfig = EventWriterConfig.builder()
+                    .transactionTimeoutTime(txnTimeoutMillis)
+                    .transactionTimeoutScaleGracePeriod(txnGracePeriodMillis)
+                    .build();
+            pravegaWriter = clientFactory.createEventWriter(streamName, eventSerializer, writerConfig);
+        }
+
+        abstract void open() throws Exception;
+
+        abstract void write(T event) throws Exception;
+
+        void close() throws Exception {
+            pravegaWriter.close();
+        }
+
+        abstract List<UUID> snapshotState(long checkpointId, long checkpointTime) throws Exception;
+
+        abstract void restoreState(List<UUID> list) throws Exception;
+
+        abstract void notifyCheckpointComplete(long checkpointId) throws Exception;
+
+    }
+
+    /*
+     * Implements an {@link AbstractInternalWriter} using transactional writes to Pravega.
+     */
+    private class TransactionalWriter extends AbstractInternalWriter {
+
+        /** The currently running transaction to which we write */
+        private Transaction<T> currentTxn;
+
+        /** The transactions that are complete from Flink's view (their checkpoint was triggered),
+         * but not fully committed, because their corresponding checkpoint is not yet confirmed */
+        private final ArrayDeque<TransactionAndCheckpoint<T>> txnsPendingCommit;
+
+        TransactionalWriter(ClientFactory clientFactory) {
+            super(clientFactory);
+            this.txnsPendingCommit = new ArrayDeque<>();
+        }
+
+        @Override
+        public void open() throws Exception {
+            // start the transaction that will hold the elements till the first checkpoint
+            this.currentTxn = pravegaWriter.beginTxn();
+            log.debug("{} - started first transaction '{}'", name(), this.currentTxn.getTxnId());
+        }
+
+        @Override
+        public void write(T event) throws Exception {
+            this.currentTxn.writeEvent(eventRouter.getRoutingKey(event), event);
```

Comment:
The higher-level lifecycle of a sink function makes it impossible that it would be null, but I agree that an assertion would be handy and will add one.   The transaction cannot be created any earlier.